Documentation/design-docs/PinnedHeap.md

```diff
@@ -0,0 +1,160 @@
+2) We allow migrating non pinned objects to the pinned heap at the time they need to be pinned.
+
+1) is a much cleaner solution. Of course it also imposes a limitation which is the user needs to have control over the objects they want to pin and is ok with pinning it at allocation time. This is not always the case. However, we believe that this feature is the most important for framework components which often take care of the underlying pinning and they are in general ok with this limitation so we will go with option 1). If we ever decide that it’s important to do 2) we can add it in the future.
+
+An object allocated on the pinned heap can be referenced by other objects normally. It’s pinned by the virtue of being on the pinned heap so we do not distinguish between a pinned reference (like a pinned GC handle) and a non pinned reference. This means as long as the object is still referenced, it will stay on the pinned heap and will not be moved.
+
+## API - allocating an array on the pinned heap
+
+For users who want to allocate their objects pinned, we provide [a new API](https://github.com/dotnet/corefx/issues/31787) to allocate such an object. The API declaration is the following:
+
+```csharp
+class GC
+{
+    // generation: -1 means to let GC decide (equivalent to regular new T[])
+    // 0 means to allocate in gen0
+    // GC.MaxGeneration means to allocate in the oldest generation
+    //
+    // pinned: true means you want to pin this object that you are allocating
+    // otherwise it's not pinned.
+    //
+    // alignment: only supported if pinned is true.
+    // -1 means you don't care which means it'll use the default alignment.
+    // otherwise specify a power of 2 value that's >= pointer size
+    // the beginning of the payload of the object (&array[0]) will be aligned with 
+    // this alignment.
+    static T[] AllocateArray<T>(int length, 
+                                int generation = -1, 
+                                bool pinned = false, 
+                                int alignment = -1);
+}
+```
+
+Note that for pinned heap we are only implementing the relevant part of the API, meaning only 2 arguments will be interpreted – length and pinned. Since this is a user facing feature we impose the following limitations on the objects allocated by this API which will make the implementation cleaner and the perf better while not hindering the majority of the usage:
+
+1) As the name suggests, this API only allows array allocations. Allowing to allocate a non array object is certainly possible but at this time we do not see much value in it so for the first version of this feature we will only allow to allocate arrays. 
+
+2) The array cannot contain any references. An object has to be of a blittable type when referenced by a pinned GC handle which means the majority of our usage of pinning is already on blittable types. This means we do not need to trace objects on the pinned heap which is a nice perf characteristic.
```

Comment:
Nit: Cannot contain any references is not the same thing as blittable. For example, `char` is not blittable type so it would not be possible to allocate `char[]` on the pinned heap.

Reply:
yes, I just meant we are not limiting the usage because what we allow is a superset of what you can do that. I can make this clear.